### ISSUE TITLE
Negative Linear Regression Loss?

### ISSUE THREAD

**Issue by zmonoid** [dmlc/mxnet#2719]
Hi,

I am trying to implement DQN in MXNet,

The symbol graph I designed is like this:

```python
def get_symbol(num_actions = 20):
    data = mx.sym.Variable("data")
    target = mx.sym.Variable("target")
    mask = mx.sym.Variable("mask")
    conv1 = mx.sym.Convolution(data=data, num_filter=32, stride=(4,4),\
                               kernel=(8,8), name="conv1")
    relu1 = mx.sym.Activation(data=conv1, act_type='relu', name="relu1")
    conv2 = mx.sym.Convolution(data=relu1, num_filter=64, stride=(2,2),\
                              kernel=(4,4), name="conv2")
    relu2 = mx.sym.Activation(data=conv2, act_type='relu', name="relu2")
    conv3 = mx.sym.Convolution(data=relu2, num_filter=64, stride=(1,1),\
                              kernel=(3,3), name="conv3")
    relu3 = mx.sym.Activation(data=conv3, act_type='relu', name="relu3")
    fc4 = mx.sym.FullyConnected(data=relu3, name="fc4", num_hidden=512)
    relu4 = mx.sym.Activation(data=fc4, act_type='relu', name="relu4")
    fc5 = mx.sym.FullyConnected(data=relu4, name="fc5", num_hidden=num_actions)
    q_sa = mx.sym.sum_axis(fc5*mask, axis=1, name='q_sa')
    loss = mx.sym.LinearRegressionOutput(data=q_sa, label=target, name='linreg1')
    return fc5, loss
```

When I feed into the data, forward, and print the loss output, it shows many negative values, how come?
```python
data_shape = (32, 4, 84, 84)
net, loss = get_symbol(6)
loss_exe = loss.simple_bind(ctx=ctx, data=data_shape, grad_req="write")
loss_exe.arg_dict['data'] = st
loss_exe.arg_dict['mask'] = atx
loss_exe.arg_dict['target'] = target
loss_exe.forward()
loss_exe.backward()
loss_exe.outputs[0].asnumpy()

array([  2.52259430e-03,  -1.71138032e-04,   2.70909048e-03,
        -2.14131060e-03,   1.47889601e-04,  -1.06083439e-03,
        -1.07564800e-03,   2.23463075e-03,   2.02975608e-03,
         1.77949620e-03,   1.97678898e-03,  -1.90932793e-03,
         5.89879521e-04,  -1.73512404e-03,  -8.45411851e-04,
         5.04227239e-04,   1.36505882e-03,   1.79292960e-03,
        -7.64497148e-04,  -1.49970304e-03,  -3.97288473e-04,
         1.05695380e-03,  -4.67159320e-04,  -1.16364099e-04,
        -2.51413067e-03,  -3.43617576e-05,   6.04484521e-05,
        -1.21897203e-03,   3.26335733e-03,   3.52809671e-04,
        -2.61261780e-03,   7.06866907e-04], dtype=float32)
```

**Comment by sxjscience** [comment]
@zmonoid The linear layer `fc5` can contain negative weights.

**Comment by zmonoid** [comment]
@sxjscience  Thanks for your answering.

But shall the LinearRegressionLoss be calculated as mean squared error like (||y - y_target||^2)/N ?

**Comment by sxjscience** [comment]
@zmonoid The `LinearRegressionOutput` will output the same vector as the input while backpropagate the gradient of the squared loss.

**Comment by zmonoid** [comment]
@sxjscience Thanks, but how to export the MSE loss in this case?

**Comment by sxjscience** [comment]
You can compute it after getting the forward result.

**Comment by zmonoid** [comment]
@sxjscience The result seems to be very strange if I plot the MSE this way:

```
N = 10000
loss = np.zeros(N)
for t in tqdm(range(N)):
    st, at, rt, pt, tt = RM.sample()
    executor.arg_dict['data'] = pt
    executor.forward()
    qq = mx.nd.max_axis(executor.outputs[0], axis=1).asnumpy()
    target = rt + (1 - tt) * gamma * qq
    atx = np.zeros((32,action_dim))
    for idx, val in enumerate(at):
        atx[idx, val] = 1
    # rwd = mx.nd.array(rt, ctx)
    # tmn = mx.nd.array(tt, ctx)
    # data = mx.nd.array(st, ctx)
    # target = rwd + (1 - terminal) * gamma * cc
    # mask = mx.nd.array(atx, ctx)
    loss_exe.arg_dict['data'] = st
    loss_exe.arg_dict['mask'] = atx
    loss_exe.arg_dict['target'] = target
    loss_exe.forward()
    loss_exe.backward()
    loss[t] = np.sum(np.abs(loss_exe.outputs[0].asnumpy() - target))
    for key in keys:
        SGD(key, loss_exe.arg_dict[key], loss_exe.grad_dict[key])
plt.plot(loss)
```
![image](https://cloud.githubusercontent.com/assets/16687307/16873194/35b91596-4ac4-11e6-8d96-6b06a3885e5c.png)

**Comment by sxjscience** [comment]
You can check the gradient.

**Comment by zmonoid** [comment]
@sxjscience 
I am trying to compare the same approch to train DQN using tensorflow.

```python
def build_dqn(num_actions):
    """
    Building a DQN.
    """
    inputs = tf.placeholder(tf.float32, [None, 4, 84, 84])
    
    # Inputs shape: [batch, channel, height, width] need to be changed into
    # shape [batch, height, width, channel]
    net = tf.transpose(inputs, [0, 2, 3, 1])
    net = tflearn.conv_2d(net, 32, 8, strides=4, activation='relu')
    net = tflearn.conv_2d(net, 64, 4, strides=2, activation='relu')
    net = tflearn.conv_2d(net, 64, 3, strides=1, activation='relu')
    net = tflearn.fully_connected(net, 512, activation='relu')
    q_values = tflearn.fully_connected(net, num_actions)
    return inputs, q_values

def build_graph(num_actions):
    # Create shared deep q network
    s, q_values = build_dqn(num_actions=num_actions)
    network_params = tf.trainable_variables()

    a = tf.placeholder("float", [None, num_actions])
    y = tf.placeholder("float", [None])
    action_q_values = tf.reduce_sum(tf.mul(q_values, a), reduction_indices=1)
    cost = tflearn.mean_square(action_q_values, y)
    optimizer = tf.train.RMSPropOptimizer(learning_rate)
    grad_update = optimizer.minimize(cost, var_list=network_params)

    graph_ops = {"s": s,
                 "q_values": q_values,
                 "a": a,
                 "y": y,
                 "grad_update": grad_update,
                 "cost": cost}

    return graph_ops

config = tf.ConfigProto()
config.gpu_options.allow_growth = True
with tf.Session(config=config) as session:
    graph_ops = build_graph(action_dim)
    session.run(tf.initialize_all_variables())
    s = graph_ops["s"]
    q_values = graph_ops["q_values"]
    a = graph_ops["a"]
    y = graph_ops["y"]
    grad_update = graph_ops["grad_update"]
    cost = graph_ops["cost"]
    
    N = 10000
    loss = np.zeros(N)
    for t in tqdm(range(N)):
        st, at, rt, pt, tt = RM.sample()
        qq = q_values.eval(session=session, feed_dict={s:pt})
        target = rt + (1 - tt) * gamma * np.max(qq, axis=1)
        atx = np.zeros((32,action_dim), dtype=float)
        for idx, val in enumerate(at):
            atx[idx, val] = 1
        #print action_q_values.eval(session=session, feed_dict={a:atx, inputs:st}).shape
        #print cost.eval(session=session, feed_dict={y:target, a:atx, inputs:st})
        session.run(grad_update, feed_dict={y: target, a:atx, s: st})
        loss[t]=np.mean(cost.eval(session=session, feed_dict={y: target, a:atx, s: st}))
    pylab.plot(loss)
```
Which gives more reasonable loss:
![image](https://cloud.githubusercontent.com/assets/16687307/16876638/3b5d9ae0-4ad7-11e6-9056-ae001bec87ab.png)

**Comment by zmonoid** [comment]
@sxjscience However, if I take the range 8000:10000, it will give similar result as mxnet, but the loss is much smaller. Strange!

pylab.plot(loss[8000:])
![image](https://cloud.githubusercontent.com/assets/16687307/16876733/a5e629c2-4ad7-11e6-8d8c-2391085d5444.png)

**Comment by sxjscience** [comment]
@zmonoid DQN is quite sensitive to the initialization and learning rates. You try to tune these hyper-parameters.

**Comment by zmonoid** [comment]
@sxjscience The reason is that, most of samples have **ZERO** reward, which will bring down the gradient to zeros once it fits the zero reward transitions. 

May I know how did you resolve such problems? Could I have a look at your project of DQN with MXNET?

**Comment by sxjscience** [comment]
@zmonoid Let's communicate offline for this. Do you have an email?

**Comment by zmonoid** [comment]
@sxjscience Thanks for your prompt reply, I have emailed you just now~
